Python/jit.c

```diff
@@ -0,0 +1,377 @@
+#ifdef _Py_JIT
+
+#include "Python.h"
+
+#include "pycore_abstract.h"
+#include "pycore_call.h"
+#include "pycore_ceval.h"
+#include "pycore_dict.h"
+#include "pycore_intrinsics.h"
+#include "pycore_long.h"
+#include "pycore_opcode_metadata.h"
+#include "pycore_opcode_utils.h"
+#include "pycore_optimizer.h"
+#include "pycore_pyerrors.h"
+#include "pycore_setobject.h"
+#include "pycore_sliceobject.h"
+#include "pycore_jit.h"
+
+#include "jit_stencils.h"
+
+// Memory management stuff: ////////////////////////////////////////////////////
+
+#ifndef MS_WINDOWS
+    #include <sys/mman.h>
+#endif
+
+static size_t
+get_page_size(void)
+{
+#ifdef MS_WINDOWS
+    SYSTEM_INFO si;
+    GetSystemInfo(&si);
+    return si.dwPageSize;
+#else
+    return sysconf(_SC_PAGESIZE);
+#endif
+}
+
+static void
+jit_warn(const char *message)
+{
+#ifdef MS_WINDOWS
+    int hint = GetLastError();
+#else
+    int hint = errno;
+#endif
+    PyErr_WarnFormat(PyExc_RuntimeWarning, 0, "JIT %s (%d)", message, hint);
+}
+
+static char *
+jit_alloc(size_t size)
+{
+    assert(size);
+    assert(size % get_page_size() == 0);
+#ifdef MS_WINDOWS
+    int flags = MEM_COMMIT | MEM_RESERVE;
+    char *memory = VirtualAlloc(NULL, size, flags, PAGE_READWRITE);
+    int failed = memory == NULL;
+#else
+    int flags = MAP_ANONYMOUS | MAP_PRIVATE;
+    char *memory = mmap(NULL, size, PROT_READ | PROT_WRITE, flags, -1, 0);
```

Comment:
Apple Silicon requires an extra flag for mmap (MAP_JIT) and a kernel call (see [docs](https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon#Disable-Write-Protections-Before-You-Generate-Instructions)) or is the `mprotect` call below abstracting this?

Reply:
I originally added that since I figured it was required. Then, later, I removed it just to see how it failed. Lo and behold, everything works fine. I think my assumption at the time was that the `MAP_JIT` flag was only useful for code that needed to self-modify during execution (in our case, we write once, then never modify the code again after it begins executing). I can't really tell from the docs.

Could be I'm overlooking some edge case (maybe in the presence of multiple threads?) or configuration that I don't have locally, but I figured I'd wait until it actually broke for somebody before introducing more `#ifdef`s. :)

Reply:
Fair. macOS's security controls are more lax when you're running a binary that you've compiled yourself, so this might only reappear when testing a packaged and signed version of Python

Reply:
If you can point me towards instructions for building a packaged and signed version of Python, I can try it out. I've just never done it before.

Reply:
So, on a local M2 Mac, trying to map memory with both `PROT_WRITE` *and* `PROT_EXEC` protections requires the `MAP_JIT` flag (it fails with `Permission denied` otherwise). The write also fails with `bus error` if `pthread_jit_write_protect_np` is not called before and after the write, as your linked docs show.

...but, mapping as `PROT_WRITE` and then *later* protecting as `PROT_EXEC` (as we do here) works fine without either the `MAP_JIT` flag or the `pthread_write_protect_np` calls. So that supports the idea that `MAP_JIT` is only needed when writing to memory that is currently executable, as in self-modifying code.

(Of course, none of this is on signed code, so it may in fact be required. But at this point I'm convinced it's not necessary in this case, and if we start seeing errors in signed releases we'll know that this is probably the culprit.

Reply:
For signed code we'd have to add a new entitlement, see https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_allow-jit?language=objc.  "We" being Ned, AFAIK the machinery for signing the binaries and installers is currently not in the repository and Ned is working on cleaning that up.

Apple also has some documentation on how to implement JITs, see https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon?language=objc.  This document amongst others notes that Apple enforces W^X on arm64.